bpf/bpf_sock.c

```diff
@@ -577,6 +622,67 @@ static __always_inline int sock6_xlate_v4_in_v6(struct bpf_sock_addr *ctx)
 	return -ENXIO;
 }
 
+#if defined ENABLE_NODEPORT || defined ENABLE_EXTERNAL_IP
+static __always_inline int sock6_post_bind_v4_in_v6(struct bpf_sock *ctx)
+{
+#ifdef ENABLE_IPV4
+	struct bpf_sock fake_ctx;
+	union v6addr addr6;
+
+	ctx_get_v6_src_address(ctx, &addr6);
+	if (!is_v4_in_v6(&addr6))
+		return 0;
+
+	__builtin_memset(&fake_ctx, 0, sizeof(fake_ctx));
+	fake_ctx.protocol = ctx->protocol;
+	fake_ctx.src_ip4  = addr6.p4;
+	fake_ctx.src_port = ctx->src_port;
+
+	return __sock4_post_bind(&fake_ctx);
+#endif /* ENABLE_IPV4 */
+	return 0;
+}
+
+static __always_inline int __sock6_post_bind(struct bpf_sock *ctx)
+{
+	struct lb6_service *svc;
+	struct lb6_key key = {
+		.dport		= bpf_htons(ctx->src_port),
+	};
+
+	if (!sock_proto_enabled(ctx->protocol))
+		return SYS_PROCEED;
+
+	ctx_get_v6_src_address(ctx, &key.address);
+
+	svc = __lb6_lookup_service(&key);
+	if (!svc) {
+		ctx_get_v6_src_address(ctx, &key.address);
+		key.dport = bpf_htons(ctx->src_port);
+		svc = sock6_nodeport_wildcard_lookup(&key, false);
+		if (!svc)
+			return sock6_post_bind_v4_in_v6(ctx);
+	}
+
+	if (svc && (lb6_svc_is_nodeport(svc) || lb6_svc_is_external_ip(svc))) {
+		return SYS_REJECT;
```

Comment:
This is not fully correct; needs to be `return -EADDRINUSE` since otherwise `sock6_post_bind ()` will never trigger the `SYS_REJECT`. Lets also make the other return codes in here similar to the v4 variant of your code.

Reply:
:scream: Good catch!